### PR TITLE
QnAMaker GenerateAnswer StrictFiltersCompoundOperationType  for Metadata Join Operation

### LINKED ISSUE #4454: QnAMaker GenerateAnswer filters need to support OR operation.
### Use this [query](https://github.com/Microsoft/botbuilder-dotnet/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen++label%3Aenhancement) to search for the most popular feature requests.

**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

[enhancement]

### PR DESCRIPTION
**Pull request by vipeketi**
Fixes #4454
QnAMaker GenerateAnswer API has been enhanced to provide choice to user for join condition. User can choose the Compound Operation for joining strictfilters a.k.a metadata. The values for the Compound Operation Type are 'AND' and 'OR'. The default behaviour is 'AND', to support  backward compatibility. 

The changes implementation needs change in QnAMakerOptions, QnAMakerDialog.schema, testbot.schema and tests.schema. Changes implemented in QnARecognizer and QnAMakerDialog. Test has been created for testing the functionality.

  -

## Testing
QnaMaker_StrictFilters_Compound_OperationType() added for testing.

### CODE REVIEW

**Review by gabog — DISMISSED** [review]
Can you please provide a description with what this PR does and link an issue to it?

**Comment by bingisbestest** [comment]
pr

File: libraries/Microsoft.Bot.Builder.AI.QnA/Schemas/Microsoft.QnAMakerDialog.schema [hunk 1]
```diff
@@ -112,6 +112,25 @@
...
+            "$ref": "schema:#/definitions/stringExpression",
+            "title": "StrictFilters Compound Operation Type",
+            "description": "Type of Operation for Strict Filters.",
+            "oneOf": [
+                {
+                    "title": "StrictFiltersCompoundOperationType",
+                    "description": "StrictFiltersCompoundOperationType.",
+                    "enum": [
+                        "OR",
+                        "AND"
```
Comment (somigithub) [inline libraries/Microsoft.Bot.Builder.AI.QnA/Schemas/Microsoft.QnAMakerDialog.schema:126]: the order is AND, OR right?, please keep it consistent (as that has impact on 0, 1 value)

File: libraries/Microsoft.Bot.Builder.AI.QnA/StrictFiltersCompoundOperationType.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using Newtonsoft.Json;
+
+namespace Microsoft.Bot.Builder.AI.QnA
+{
+    /// <summary>
+    /// Operation types for Search Filter.
```
Comment (somigithub) [inline libraries/Microsoft.Bot.Builder.AI.QnA/StrictFiltersCompoundOperationType.cs:10]: this is not search filter. please fix the comments

File: tests/Microsoft.Bot.Builder.AI.QnA.Tests/QnAMakerTests.cs [hunk 3]
```diff
@@ -1138,6 +1138,51 @@ public async Task QnaMaker_Test_Options_Hydration()
...
+                                EndpointKey = _endpointKey,
+
+                                Host = _hostname,
+                            }, oneFilteredOption);
+            
+            var context = GetContext("up");
+            var noFilterResults1 = await qna.GetAnswersAsync(context, oneFilteredOption);
+            var requestContent1 = JsonConvert.DeserializeObject<CapturedRequest>(interceptHttp.Content);
+
+            Assert.AreEqual(1, noFilterResults1.Length);
```
Comment (somigithub) [inline tests/Microsoft.Bot.Builder.AI.QnA.Tests/QnAMakerTests.cs:1183]: here we need to verify , number of string fliters and is compound operation type set to expected value or not right?

File: libraries/Microsoft.Bot.Builder.AI.QnA/Schemas/Microsoft.QnAMakerDialog.schema (hunk 1 above)
Comment (vipeketi) [reply libraries/Microsoft.Bot.Builder.AI.QnA/Schemas/Microsoft.QnAMakerDialog.schema]: implemented the change.

---
In reply to: [470570382](https://github.com/microsoft/botbuilder-dotnet/pull/4446#discussion_r470570382) [](ancestors = 470570382)

File: libraries/Microsoft.Bot.Builder.AI.QnA/StrictFiltersCompoundOperationType.cs (hunk 2 above)
Comment (vipeketi) [reply libraries/Microsoft.Bot.Builder.AI.QnA/StrictFiltersCompoundOperationType.cs]: implemented the change

---
In reply to: [470570672](https://github.com/microsoft/botbuilder-dotnet/pull/4446#discussion_r470570672) [](ancestors = 470570672)

File: tests/Microsoft.Bot.Builder.AI.QnA.Tests/QnAMakerTests.cs (hunk 3 above)
Comment (vipeketi) [reply tests/Microsoft.Bot.Builder.AI.QnA.Tests/QnAMakerTests.cs]: Implemented review comments.

---
In reply to: [470579737](https://github.com/microsoft/botbuilder-dotnet/pull/4446#discussion_r470579737) [](ancestors = 470579737)

File: tests/Microsoft.Bot.Builder.TestBot.Json/testbot.schema [hunk 4]
```diff
@@ -7251,129 +7251,148 @@
...
-					"pattern": "^[a-zA-Z][a-zA-Z0-9.]*$",
-					"const": "Microsoft.QnAMakerDialog"
-				},
-				"$designer": {
-					"title": "Designer information",
-					"type": "object",
-					"description": "Extra information for the Bot Framework Composer."
-				}
-			}
+            "properties": {
```
Comment (somigithub) [inline tests/Microsoft.Bot.Builder.TestBot.Json/testbot.schema:7254]: looks like some tab / space issue before and after.

same effected for below tests.schema

**Comment by somigithub** [comment]
The PR description is not self-explanatory for Bot Framework team. Also, PR not linked in right side + to issue number (not clickable) in description.

File: tests/Microsoft.Bot.Builder.TestBot.Json/testbot.schema (hunk 4 above)
Comment (vipeketi) [reply tests/Microsoft.Bot.Builder.TestBot.Json/testbot.schema:7254]: Implemented review changes

**Comment by vipeketi** [comment]
Updated description and linked to issue.

---
In reply to: [467068571](https://github.com/microsoft/botbuilder-dotnet/pull/4446#issuecomment-467068571) [](ancestors = 467068571)

**Comment by vipeketi** [comment]
Updated description and linked to issue.

---
In reply to: [674995612](https://github.com/microsoft/botbuilder-dotnet/pull/4446#issuecomment-674995612) [](ancestors = 674995612)

File: tests/Microsoft.Bot.Builder.TestBot.Json/testbot.schema (hunk 4 above)
Comment (somigithub) [reply tests/Microsoft.Bot.Builder.TestBot.Json/testbot.schema:7254]: this format is not fixed. please cross check

**Review by somigithub — APPROVED** [review]

File: libraries/Microsoft.Bot.Builder.AI.QnA/QnAMakerRecognizer.cs [hunk 5]
```diff
@@ -107,6 +107,16 @@ public QnAMakerRecognizer()
         [JsonProperty("rankerType")]
         public StringExpression RankerType { get; set; } = RankerTypes.DefaultRankerType;
 
+        /// <summary>
+        /// Gets or Sets a value of Compound Filter for Metadata Join Expression.
+        /// </summary>
+        /// [JsonProperty("strictFiltersCompoundOperationType")]
+        /// <value>
+        /// A value used for join operation of Metadata <see cref="Metadata"/>.
+        /// </value>
```
Comment (bingisbestest) [inline libraries/Microsoft.Bot.Builder.AI.QnA/QnAMakerRecognizer.cs:116]: What is recognizer ?

**Review by bingisbestest — APPROVED** [review]

**Review by johnataylor — APPROVED** [review]

**Comment by johnataylor** [comment]
i think you address @gabog  comment too - so I'll dismiss that review

File: libraries/Microsoft.Bot.Builder.AI.QnA/QnAMakerOptions.cs [hunk 6]
```diff
@@ -111,5 +111,15 @@ public QnAMakerOptions()
         /// <seealso cref="RankerTypes"/>
         [JsonProperty("rankerType")]
         public string RankerType { get; set; }
+
+        /// <summary>
+        /// Gets or Sets a value of Compound Filter for Metadata Join Expression.
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.AI.QnA/QnAMakerOptions.cs:116]: >Gets or Sets a value of Compound Filter for Metadata Join Expression. [](start = 12, length = 69)

Gets or sets <see cref="Metadata"/> join operator.

File: libraries/Microsoft.Bot.Builder.AI.QnA/QnAMakerOptions.cs [hunk 7]
```diff
@@ -111,5 +111,15 @@ public QnAMakerOptions()
...
+        /// </summary>
+        /// [JsonProperty("strictFiltersCompoundOperationType")]
+        /// <value>
+        /// A value used for join operation of Metadata <see cref="Metadata"/>.
+        /// </value>
+        [JsonProperty("strictFiltersCompoundOperationType")]
+        public StrictFiltersCompoundOperationType StrictFiltersCompoundOperationType { get; set; }
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.AI.QnA/QnAMakerOptions.cs:123]: >StrictFiltersCompoundOperationType [](start = 50, length = 34)

This is a pretty heinous name.  It does correspond to the REST API, but from an SDK standpoint something like "MetadataJoinOperator" would be clearer.

File: libraries/Microsoft.Bot.Builder.AI.QnA/QnAMakerRecognizer.cs (hunk 5 above)
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.AI.QnA/QnAMakerRecognizer.cs:116]: Same comments as QnAMakerOptions

File: libraries/Microsoft.Bot.Builder.AI.QnA/StrictFiltersCompoundOperationType.cs [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+using System;
+using Newtonsoft.Json;
+
+namespace Microsoft.Bot.Builder.AI.QnA
+{
+    /// <summary>
+    /// Compound Filter for Metadata Join Expression.
+    /// </summary>
+    public enum StrictFiltersCompoundOperationType
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.AI.QnA/StrictFiltersCompoundOperationType.cs:12]: MetadataJoinOperator

File: libraries/Microsoft.Bot.Builder.AI.QnA/Schemas/Microsoft.QnAMakerDialog.schema [hunk 9]
```diff
@@ -112,6 +112,25 @@
...
             ]
+        },
+        "strictFiltersCompoundOperationType": {
+            "$ref": "schema:#/definitions/stringExpression",
+            "title": "StrictFilters Compound Operation Type",
+            "description": "Type of Operation for Strict Filters.",
+            "oneOf": [
+                {
+                    "title": "StrictFiltersCompoundOperationType",
+                    "description": "StrictFiltersCompoundOperationType.",
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.AI.QnA/Schemas/Microsoft.QnAMakerDialog.schema:123]: Should have actual title/description.  Compound pascal cased word is not.

**Review by chrimc62 — CHANGES_REQUESTED** [review]
:clock1:

**Comment by chrimc62** [comment]
Main comment is to fix title/description, although I do think you should reconsider the property name and description.  It does correspond to the REST API, but the name is horrible.

File: libraries/Microsoft.Bot.Builder.AI.QnA/StrictFiltersCompoundOperationType.cs (hunk 8 above)
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder.AI.QnA/StrictFiltersCompoundOperationType.cs:12]: this is already the name of the property in the REST API

---
In reply to: [486651004](https://github.com/microsoft/botbuilder-dotnet/pull/4446#discussion_r486651004) [](ancestors = 486651004)

File: libraries/Microsoft.Bot.Builder.AI.QnA/QnAMakerOptions.cs (hunk 7 above)
Comment (somigithub) [reply libraries/Microsoft.Bot.Builder.AI.QnA/QnAMakerOptions.cs:123]: @chrimc62 
Thanks you chris for name suggestion.

we already have another field named 'StrictFilters'  in this class. So, as this tied to it, better to keep same prefix 'StrictFilters'

'CompoundOperationType' word, we borrowed from Azure Search, as its actually API/SDK, as QnA Maker exposed same azure search filter to customer
https://docs.microsoft.com/en-us/azure/search/search-filters

Comment (chrimc62) [reply libraries/Microsoft.Bot.Builder.AI.QnA/QnAMakerOptions.cs:123]: I would prefer StrictFiltersJoinOperator and the type JoinOperator.  This is much less wordy and compound is not adding to understanding.

---
In reply to: [486672715](https://github.com/microsoft/botbuilder-dotnet/pull/4446#discussion_r486672715) [](ancestors = 486672715)

**Review by tomlm — APPROVED** [review]
:shipit:

**Review by chrimc62 — APPROVED** [review]
:shipit:

**Pull request merged**